src/Microsoft.Diagnostics.Runtime/src/Common/ClrArray.cs

```diff
@@ -157,15 +157,65 @@ public int GetUpperBound(int dimension)
 
         public T GetValue<T>(params int[] indices) where T : unmanaged => GetValue<T>(false, indices);
 
+        // This helper method returns the address of the item at a given index of an array.
+        // - reference type object: the array item at the given index contains the address so it has to be read.
+        // - value type object: it will be embedded inside the array itself so the address is computed relatively to the array itself.
+        private ulong GetItemAddress(int index, bool isObjectReference)
+        {
+            if (Rank != 1)
+                throw new ArgumentException($"Array {Address:x} was not a one-dimensional array.  Type: {Type?.Name ?? "null"}");
+
+            int valueOffset = index;
+            int dataByteOffset = 2 * IntPtr.Size;
+
+            if (IsMultiDimensional)
+            {
+                valueOffset -= GetMultiDimensionalBound(1);
+                if ((uint)valueOffset >= GetMultiDimensionalBound(0))
+                    throw new ArgumentOutOfRangeException(nameof(index));
+
+                dataByteOffset += 2 * sizeof(int);
+            }
+            else
+            {
+                if ((uint)valueOffset >= Length)
+                    throw new ArgumentOutOfRangeException(nameof(index));
+            }
+
+
+            int elementSize = isObjectReference ? UIntPtr.Size : Type.ComponentSize;
+            int valueByteOffset = dataByteOffset + valueOffset * elementSize;
+
+            if (isObjectReference)
+            {
+                return Type.ClrObjectHelpers.DataReader.Read<UIntPtr>(Address + (ulong)valueByteOffset).ToUInt64();
+            }
+            else
+            {
+                return Address + (ulong)valueByteOffset;
+            }
+        }
+
+        public ClrValueType GetStructValue(int index)
+        {
+            if (Type.ComponentType.IsObjectReference)
```

Comment:
ComponentType needs to be null checked.  Unfortunately there's some types we just can't get information on, and an array of generic types sometimes has a null ComponentType.  I think returning an empty ClrValueType (which will has `ClrValueType.IsValid == false`) is the only sensible thing to do in that case.

Reply:
Good idea.